dpll_mon.c

```diff
@@ -0,0 +1,991 @@
+		return;
+	pin_id = nla_get_u32(tb[DPLL_A_PIN_ID]);
+
+	nla_for_each_attr(a, genlmsg_attrdata(gnlh, 0),
+			  genlmsg_attrlen(gnlh, 0), rem) {
+		if (nla_type(a) == DPLL_A_PIN_PARENT_DEVICE)
+			update_pin(dm, pin_id, a, exist, notify);
+		else if (nla_type(a) == DPLL_A_PIN_PARENT_PIN)
+			update_muxed_pin(dm, pin_id, a, exist, notify);
+	}
+}
+
+static void dpll_mon_device_update(struct dpll_mon *dm, struct nlattr **tb)
+{
+	dpll_mon_device_id_update(dm, tb);
+	if (tb[DPLL_A_MODE])
+		dm->dpll_mode = nla_get_u32(tb[DPLL_A_MODE]);
+	if (tb[DPLL_A_LOCK_STATUS])
+		dm->lock_status = nla_get_u32(tb[DPLL_A_LOCK_STATUS]);
+}
+
+static int dpll_mon_device_owner(struct dpll_mon *dm, struct nlattr **tb)
+{
+	if (tb[DPLL_A_CLOCK_ID] && tb[DPLL_A_MODULE_NAME] && tb[DPLL_A_TYPE])
+		if (nla_get_u64(tb[DPLL_A_CLOCK_ID]) == dm->clock_id &&
+		    !nla_strcmp(tb[DPLL_A_MODULE_NAME],
+				dm->module_name) &&
+		    nla_get_u32(tb[DPLL_A_TYPE]) == DPLL_TYPE_EEC)
+			return 1;
+
+	return 0;
+}
+
+static int pins_ready(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list) {
+		if (!pin->ready) {
+			dm->init_err++;
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+static int pins_dnu(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list) {
+		if (pin->prio_valid && pin->prio != dm->dev_dnu_prio) {
+			pr_debug("not expected pin id:%u prio: %u",
+				 pin->id, pin->prio);
+			dm->init_err++;
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+static void invalidate_pins(struct dpll_mon *dm)
+{
+	struct dpll_mon_pin *pin;
+
+	STAILQ_FOREACH(pin, &dm->pins, list)
+		pin->ready = 0;
+}
+
+static int dpll_mon_ntf_recv(struct nl_msg *msg, void *arg)
+{
+	struct nlattr *tb[MAX_ATTR + 1];
+	struct dpll_mon *dm = arg;
+	struct genlmsghdr *gnlh;
+	int ret;
+
+	gnlh = nlmsg_data(nlmsg_hdr(msg));
+	if (!gnlh) {
+		pr_err("failed parse gnl header");
+		return -EINVAL;
+	}
+	ret = genlmsg_parse(nlmsg_hdr(msg), 0, tb, MAX_ATTR, NULL);
+	if (ret) {
+		pr_err("genlsmg_parse err:%d", ret);
+		return ret;
+	}
+	switch (gnlh->cmd) {
+	case DPLL_CMD_DEVICE_CREATE_NTF:
+		if ((dm->state == DPLL_MON_STATE_INIT_READY ||
+		     dm->state == DPLL_MON_STATE_DEV_GET_WAIT) &&
+		    dpll_mon_device_owner(dm, tb)) {
+			dpll_mon_device_update(dm, tb);
+			dpll_mon_state_set(dm, DPLL_MON_STATE_DEV_INITED);
+		}
+		break;
+	case DPLL_CMD_DEVICE_CHANGE_NTF:
+		if (dpll_mon_device_owner(dm, tb))
+			dpll_mon_device_update(dm, tb);
+		break;
+	case DPLL_CMD_DEVICE_DELETE_NTF:
+		if (dpll_mon_device_owner(dm, tb))
+			dpll_mon_state_set(dm, DPLL_MON_STATE_DPLL_REINIT);
+		break;
+	case DPLL_CMD_PIN_DELETE_NTF:
+		dpll_mon_pin_update(dm, tb, 0, 1, gnlh);
+		break;
+	case DPLL_CMD_PIN_CREATE_NTF:
+	case DPLL_CMD_PIN_CHANGE_NTF:
+		dpll_mon_pin_update(dm, tb, 1, 1, gnlh);
+		break;
+	default:
+		break;
+	}
+
+	return NL_OK;
+}
+
+static int dpll_rt_recv(struct nl_msg *msg, void *arg)
+{
+	struct nlattr *tb[RTM_MAX + 1], *an;
```

Comment:
Should RTM_MAX be IFLA_MAX to correspond to the nlmsg_parse() call below?

Reply:
yes, already fixed it, good catch!